pytorch_lightning/trainer/supporters.py

```diff
@@ -119,27 +122,249 @@ def mean(self):
+
+        dl_args = {
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in skip_keys
+        }
+
+        multiprocessing_context = dataloader.multiprocessing_context
+
+        # override parameters to enable batch_sampler injection
+        dl_args["batch_size"] = 1
+        dl_args["sampler"] = None
+        dl_args["shuffle"] = None
+        dl_args["batch_sampler"] = LightningBatchSamplerWrapper(dataloader.batch_sampler)
+        dl_args["drop_last"] = False
+        dl_args['multiprocessing_context'] = multiprocessing_context
+
+        missing_kwargs = valid_kwargs.difference(skip_valid_keys).difference(set(dl_args))
+        if len(missing_kwargs) != 0:
+            dataloader_cls_name = dataloader.__class__.__name__
+            rank_zero_warn(
+                f"Trying to replace to wrap your BatchSampler in your {dataloader_cls_name} dataloader."
+                "This would fail as your DataLoader doesn't expose as attributes all its __init__ parameters. "
+                f"Missing attributes are {missing_kwargs}"
+                "HINT: use Trainer(enable_predict_auto_id=False) and provide your own id. "
+                "Check out the doc for Testing. ", UserWarning
+            )
+            return dataloader
+
+        if not contains_dataset:
+            dl_args.pop('dataset')
+
+        dataloader = type(dataloader)(**dl_args)
+        dataloader.multiprocessing_context = multiprocessing_context
+        return dataloader
+
+
 class PredictionCollection(object):
-    def __init__(self, global_rank: int, world_size: int):
+
+    """
+    This class is used to collect predictions.
+
+    The legacy API built using the following functions:
+        - add
+        - _add_prediction
+        - to_disk
+
+    The new API built using the following functions:
+        - cache: Receive a new predictions to cache. Handle validation
+        - _cache_prediction: Handle storing
+        - attach_predictions: When the epoch is finished, add predictions to results object
+        - all_gather_predictions: Gather predictions accross multiple processses.
+
+    """
+
+    ID_KEY = 'id'
+
+    def __init__(self, global_rank: int, world_size: int, all_gather_fn: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather_fn = all_gather_fn
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self) -> Dict:
+        return self._predictions[self.current_stage]
+
+    def _cache_prediction(
+        self,
+        predictions: List[Dict],
+        dl_idx: int,
+        batch_indices: List[int],
+        enable_predict_auto_id: bool
+    ) -> None:
+
+        cache = self.predictions
+
+        cache.setdefault(dl_idx, {})
+
+        def convert(value):
```

Comment:
```suggestion
        def _convert(value):
```
I may make it protected so the reader does not search for it in top-level imports, but no big deal :]